modules/perf/class.jetpack-perf-optimize-assets.php

```diff
@@ -0,0 +1,279 @@
+		// 	return true;
+		// }
+
+		$registration = $wp_scripts->registered[$handle];
+
+		// inline anything local, with a src starting with /, or starting with site_url
+		$site_url = site_url();
+		// TODO: handle //, like //stats.wp.com/w.js - whoops!
+		if ( strncmp( $registration->src, '/', 1 ) === 0 ) {
+			$registration->extra['jetpack-inline'] = true;
+			$registration->extra['jetpack-inline-file'] = untrailingslashit( ABSPATH ) . $registration->src;
+		} elseif ( strpos( $registration->src, $site_url ) === 0 ) {
+			$registration->extra['jetpack-inline'] = true;
+			$raw_path = substr( $registration->src, strlen( $site_url ) );
+			$registration->extra['jetpack-inline-file'] = untrailingslashit( ABSPATH ) . $raw_path;
+		}
+
+		return isset( $registration->extra['jetpack-inline'] ) && $registration->extra['jetpack-inline'];
+	}
+
+	private function get_inline_script_content( $handle ) {
+		global $wp_scripts;
+		$registration = $wp_scripts->registered[$handle];
+
+		if ( isset( $registration->extra['jetpack-inline-file'] ) && file_exists( $registration->extra['jetpack-inline-file'] ) ) {
+			$file_path = $registration->extra['jetpack-inline-file'];
+		} else {
+			return "console.warn('failed to get script contents for " . $handle . "');";
+		}
+
+		// TODO: file_exists
+		return file_get_contents( $file_path );
+	}
+
+	/** STYLES **/
+
+	public function filter_inline_styles( $src, $handle ) {
+		// reset src to empty - can't return empty string though because then it skips rendering the tag
+		if ( $this->should_inline_style( $handle ) ) {
+			return '#';
+		}
+
+		return $src;
+	}
+
+	public function print_inline_styles( $tag, $handle, $href, $media ) {
+		if ( $this->should_inline_style( $handle ) ) {
+			return "<style type='text/css' media='$media'>" . $this->get_inline_style_content( $handle ) . '</style>';
+		}
+
+		if ( $this->should_remove_style( $handle ) ) {
+			return '';
+		}
+
+		return $tag;
+	}
+
+	private function should_inline_style( $handle ) {
+		if ( ! $this->inline_scripts_and_styles ) {
+			return false;
+		}
+
+		global $wp_styles;
+
+		if ( ! isset( $wp_styles->registered[$handle] ) ) {
+			return false;
+		}
+
+		$registration = $wp_styles->registered[$handle];
+
+		// inline anything local, with a src starting with /, or starting with site_url
+		$site_url = site_url();
+		if ( strncmp( $registration->src, '/', 1 ) === 0 ) {
+			$registration->extra['jetpack-inline'] = true;
+			$registration->extra['jetpack-inline-file'] = untrailingslashit( ABSPATH ) . $registration->src;
+		} elseif ( strpos( $registration->src, $site_url ) === 0 ) {
+			$registration->extra['jetpack-inline'] = true;
+			$raw_path = substr( $registration->src, strlen( $site_url ) );
+			$registration->extra['jetpack-inline-file'] = untrailingslashit( ABSPATH ) . $raw_path;
+		}
+
+		return isset( $registration->extra['jetpack-inline'] ) && $registration->extra['jetpack-inline'];
+	}
+
+	private function should_remove_style( $handle ) {
```

Comment:
For locally defined (inlined fonts), a better solution might be to inject a font-display [1] swap? For external resources, like Typekit and Google Fonts, that won't work today, but hopefully soon, once implementations for [2] land in Chrome and other browsers.
Alternatively, have you considered non-blocking font load / loadCSS [3] variant instead? That introduces FOUT, but I still think it's better than omitting them entirely? 

---

[1] https://www.w3.org/TR/css-fonts-4/#font-display-desc
[2] https://www.w3.org/TR/css-fonts-4/#font-display-font-feature-values
[3] https://github.com/filamentgroup/loadCSS#recommended-loadcss-usage